### PR TITLE
rsa/rsa_pmeth.c: Add the checks for the EVP_MD_CTX_get_size()

### PR DESCRIPTION
**Pull request by JiangJias**
Add the checks for the return value of EVP_MD_CTX_get_size() before explicitly cast them to size_t to avoid the integer overflow.

Fixes: 75d44c0452 ("Store digests as EVP_MD instead of a NID.")

##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/rsa/rsa_pmeth.c [hunk 1]
```diff
@@ -144,9 +144,14 @@ static int pkey_rsa_sign(EVP_PKEY_CTX *ctx, unsigned char *sig,
      * be reflected back in the "original" key.
      */
     RSA *rsa = (RSA *)EVP_PKEY_get0_RSA(ctx->pkey);
+    int md_size;
 
     if (rctx->md) {
-        if (tbslen != (size_t)EVP_MD_get_size(rctx->md)) {
+        md_size = EVP_MD_get_size(rctx->md);
+        if (md_size <= 0) {
```
Comment (slontis) [inline crypto/rsa/rsa_pmeth.c:151]: remove {} from single liners.

Comment (JiangJias) [reply crypto/rsa/rsa_pmeth.c]: Thanks. I have add a new commit to solve the problem.

File: crypto/rsa/rsa_pmeth.c [hunk 2]
```diff
@@ -144,9 +144,14 @@ static int pkey_rsa_sign(EVP_PKEY_CTX *ctx, unsigned char *sig,
      * be reflected back in the "original" key.
      */
     RSA *rsa = (RSA *)EVP_PKEY_get0_RSA(ctx->pkey);
+    int md_size;
 
     if (rctx->md) {
-        if (tbslen != (size_t)EVP_MD_get_size(rctx->md)) {
+        md_size = EVP_MD_get_size(rctx->md);
+        if (md_size <= 0)
+            return -1;
```
Comment (t8m) [inline crypto/rsa/rsa_pmeth.c:152]: Please join this with the following if as I think raising RSA_R_INVALID_DIGEST_LENGTH is appropriate here too.

File: crypto/rsa/rsa_pmeth.c [hunk 3]
```diff
@@ -266,12 +271,16 @@ static int pkey_rsa_verify(EVP_PKEY_CTX *ctx,
...
     size_t rslen;
+    int md_size;
 
     if (rctx->md) {
         if (rctx->pad_mode == RSA_PKCS1_PADDING)
             return RSA_verify(EVP_MD_get_type(rctx->md), tbs, tbslen,
                               sig, siglen, rsa);
-        if (tbslen != (size_t)EVP_MD_get_size(rctx->md)) {
+        md_size = EVP_MD_get_size(rctx->md);
+        if (md_size <= 0)
```
Comment (t8m) [inline crypto/rsa/rsa_pmeth.c:281]: Please join this with the following if as I think raising RSA_R_INVALID_DIGEST_LENGTH is appropriate here too.

File: crypto/rsa/rsa_pmeth.c [hunk 4]
```diff
@@ -485,8 +495,11 @@ static int pkey_rsa_ctrl(EVP_PKEY_CTX *ctx, int type, int p1, void *p2)
                     ERR_raise(ERR_LIB_RSA, RSA_R_INVALID_PSS_SALTLEN);
                     return -2;
                 }
+                md_size = EVP_MD_get_size(rctx->md);
+                if (md_size <= 0)
+                    return -2;
```
Comment (t8m) [inline crypto/rsa/rsa_pmeth.c:500]: I think raising RSA_R_INVALID_DIGEST_LENGTH is appropriate here too.

File: crypto/rsa/rsa_pmeth.c [hunk 5]
```diff
@@ -864,7 +877,10 @@ static int pkey_pss_init(EVP_PKEY_CTX *ctx)
         return 0;
 
     /* See if minimum salt length exceeds maximum possible */
-    max_saltlen = RSA_size(rsa) - EVP_MD_get_size(md);
+    md_size = EVP_MD_get_size(md);
+    if (md_size <= 0)
```
Comment (t8m) [inline crypto/rsa/rsa_pmeth.c:881]: I think raising RSA_R_INVALID_DIGEST_LENGTH is appropriate here too.

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/rsa/rsa_pmeth.c (hunk 5 above)
Comment (JiangJias) [reply crypto/rsa/rsa_pmeth.c]: Thanks. I have submitted a new commit to add all the needed ERR_raise().

**Review by t8m — APPROVED** [review]

File: crypto/rsa/rsa_pmeth.c [hunk 6]
```diff
@@ -864,7 +883,12 @@ static int pkey_pss_init(EVP_PKEY_CTX *ctx)
         return 0;
 
     /* See if minimum salt length exceeds maximum possible */
-    max_saltlen = RSA_size(rsa) - EVP_MD_get_size(md);
+    md_size = EVP_MD_get_size(md);
+    if (md_size <= 0) {
+        ERR_raise(ERR_LIB_RSA, RSA_R_INVALID_DIGEST_LENGTH);
+        return 0;
+    }
+    max_saltlen = RSA_size(rsa) - md_size;
```
Comment (tmshort) [inline crypto/rsa/rsa_pmeth.c:891]: Should we check for `RSA_size(rsa) < md_size` to avoid a negative `max_saltlen`?

Comment (JiangJias) [reply crypto/rsa/rsa_pmeth.c:891]: I find there is already a check afterward "if (min_saltlen > max_saltlen)." And only if "max_saltlen" is MIN_INT can an overflow occur in the "if ((RSA_bits(rsa) & 0x7) == 1) max_saltlen--;". However, subtracting two positive numbers cannot result in MIN_INT.

**Review by nhorman — APPROVED** [review]

**Comment by t8m** [comment]
Squashed and merged to the master branch. Thank you for your contribution.

**Pull request closed without merging**